pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+{
+    /* When dropping privileges from root, the `setgroups` call will
+    * remove any extraneous groups. If we don't call this, then
+    * even though our uid has dropped, we may still have groups
+    * that enable us to do super-user things. This will fail if we
+    * aren't root, so don't bother checking the return value, this
+    * is just done as an optimistic privilege dropping function.
+    */
+    setgroups(0, NULL);
+
+    //Change to the user in case we are not it yet
+    if (setgid (userInfo->pw_gid) < 0 || setuid (userInfo->pw_uid) < 0 ||
+        setegid (userInfo->pw_gid) < 0 || seteuid (userInfo->pw_uid) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static void execute_kwallet(pam_handle_t *pamh, struct passwd *userInfo, int toWalletPipe[2], char *fullSocket)
+{
+    //In the child pam_syslog does not work, using syslog directly
+
+    //keep stderr open so socket doesn't returns us that fd
+    int x = 3;
+    //Set FD_CLOEXEC on fd that are not of interest of kwallet
+    for (; x < 64; ++x) {
+        if (x != toWalletPipe[0]) {
+            fcntl(x, F_SETFD, FD_CLOEXEC);
+        }
+    }
+
+    //This is the side of the pipe PAM will send the hash to
+    close (toWalletPipe[1]);
+
+    //Change to the user in case we are not it yet
+    if (drop_privileges(userInfo) < 0) {
+        syslog(LOG_ERR, "%s: could not set gid/uid/euid/egit for kwalletd", logPrefix);
+        goto cleanup;
+    }
+
+    int envSocket;
+    if ((envSocket = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
+        syslog(LOG_ERR, "%s: couldn't create socket", logPrefix);
+        goto cleanup;
+    }
+
+    struct sockaddr_un local = {};
+    local.sun_family = AF_UNIX;
+
+    if (strlen(fullSocket) > sizeof(local.sun_path)) {
+        syslog(LOG_ERR, "%s: socket path %s too long to open",
+                   logPrefix, fullSocket);
+        free(fullSocket);
+        goto cleanup;
+    }
+    strcpy(local.sun_path, fullSocket);
+    unlink(local.sun_path);//Just in case it exists from a previous login
+
+    syslog(LOG_DEBUG, "%s: final socket path: %s", logPrefix, local.sun_path);
+
+    if (bind(envSocket, (struct sockaddr *)&local, sizeof(local)) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't bind to local file\n", logPrefix);
+        goto cleanup;
+    }
+
+    if (listen(envSocket, 5) == -1) {
+        syslog(LOG_INFO, "%s-kwalletd: Couldn't listen in socket: %d-%s\n", logPrefix, errno, strerror(errno));
+        goto cleanup;
+    }
+    //finally close stderr
+    close(2);
+
+    // Fork twice to daemonize kwallet
+    setsid();
+    pid_t pid = fork();
+    if (pid != 0) {
+        if (pid == -1) {
+            exit(EXIT_FAILURE);
+        } else {
+            exit(0);
+        }
+    }
+
+    //TODO use a pam argument for full path kwalletd
+    char pipeInt[4];
+    sprintf(pipeInt, "%d", toWalletPipe[0]);
+    char sockIn[4];
+    sprintf(sockIn, "%d", envSocket);
```

Comment:
Although typically file descriptor numbers are below 1024, it is unclean to use unbounded `sprintf()` here. This should use `snprintf()` and the result should be checked whether the output buffer was large enough.